### PR TITLE
Shader injection system (per RFC)

### PR DESCRIPTION
**Pull request by ibgreen**
For #548 
#### Background
- Very basic "templating" system for shaders
#### Change List
- Add shader inject logic

### CODE REVIEW

File: src/shadertools/src/lib/inject-shader.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+const MODULE_INJECTORS = {
+  [VERTEX_SHADER]: MODULE_INJECTORS_VS,
+  [FRAGMENT_SHADER]: MODULE_INJECTORS_FS
+};
+
+// A minimal shader injection/templating system.
+// RFC: https://github.com/uber/luma.gl/blob/master/dev-docs/RFCs/v6.0/shader-injection-rfc.md
+export default function injectShader(source, type, inject, injectStandardStubs) {
+  const INJECT_DECLARATIONS = /^/;
+  const INJECT_CODE = /}[^{}]*$/;
```
Comment (Pessimistress) [inline src/shadertools/src/lib/inject-shader.js]: - Nit: Define constants outside this function
- Is it guaranteed that main() is the last in the shader code? Add a comment about the assumption here?

File: src/shadertools/src/lib/inject-shader.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      break;
+    case 'decl-fs':
+      if (!isVertex) {
+        source = source.replace(INJECT_DECLARATIONS, fragment);
+      }
+      break;
+    // main code is injected at the end of main function
+    case 'main-vs':
+      if (isVertex) {
+        source = source.replace(INJECT_CODE, fragment.concat('\n}\n'));
```
Comment (Pessimistress) [inline src/shadertools/src/lib/inject-shader.js]: To make sure we don't accidentally delete anything:

```
source = source.replace(INJECT_CODE, match => fragment + match[0]);
```

**Comment by ibgreen** [comment]
> Nit: Define constants outside this function

done

> Is it guaranteed that main() is the last in the shader code? Add a comment about the assumption here?

done

> To make sure we don't accidentally delete anything:

done

**Comment by ibgreen** [comment]
> I'd love it if this feature can support injecting to the beginning of the main() function as well. Then it's possible to do things like:

Done. Test case updated to show this working.

**Pull request merged**